### DISCUSSION TITLE
Hackathon: testing how programs work in Shadow

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#2037] [Show and tell]
debugging tips
- for installing dbg packages
  - apt install debian-goodies
  - find-dbgsym-packages <pid>

iperf3
- segfault caused by a bug in getaddrinfo(): https://github.com/shadow/shadow/pull/1998
- must bind the server to an IPv4 address ('--bind 0.0.0.0')
- recommended to use a short time period for the client (ex: '--time 5')
- requires getsockopt and setsockopt support for TCP_CONGESTION
- after fixing the segfault and adding support for TCP_CONGESTION, iperf3 spins in the main eventloop: https://github.com/esnet/iperf/blob/dfcea9f6a09ead01089a3c9d20c7032f2c0af2c1/src/iperf_client_api.c#L487
  - this spinloop occurs both before and after the main 'TEST_RUNNING' state
  - the main loop is waiting for the 'test->ctrl_sck' to become readable, but it's also adding that control socket into the 'writefds' set, so since it will always be writable, the select() syscall will repeatedly return immediately with a rv of 1
  - iperf3 works after setting 'unblocked_syscall_limit: 500'
    - the client completes successfully, but the server returns an error code
- for the ubuntu 20.04 version of iperf3:
  - 'handle SIGSYS noprint nostop'
  - 'break iperf_client_api.c:492' for the select syscall
  - 'call sleep(2)' to escape from the spinloop (the client should send the control socket a message within these two seconds; might need to sleep a couple of times)

iperf2
- recommended to use a short time period for the client (ex: '--time 5')
- requires the 'setiterm' syscall
- will run if you build iperf without support for the 'setiterm' syscall (use '#define HAVE_SETITIMER 0')

curl (older versions)
- https://github.com/shadow/shadow/issues/1794

ngingx (works with certain config settings)
- https://github.com/shadow/shadow/pull/1179

nmap (-sT)
- the regular file (File*) flag support is broken in shadow, which causes shadow to deadlock
- might want to improve how files are handled in shadow
  - in Linux, disk-based files are always blocking (NONBLOCK has no effect)
  - shadow treats non-disk-based files as regular files, for example /dev/tty
- i think we can also work around this with the '--noninteractive' flag
- tty code: https://github.com/nmap/nmap/blob/df33da47228e3f32f9a332f7db0a0a4f2f14084d/nmap_tty.cc
- after working around the file flag issue and using the '--noninteractive' flag, nmap gets stuck in a select() loop and time never moves forward
  - nmap works after setting 'unblocked_syscall_limit: 500'

**Comment by stevenengler** [comment]
Some of these issues are now supported by shadow, and the remaining issues (except for nmap) are now reflected in the documentation. For example: https://shadow.github.io/docs/guide/iperf3.html
